Refactoring and addition of features to MySQL implementation

souris-dev opened the pull request:
The existing code for the MySQL implementation lacked certain features and could use a revamp, so I took my liberty and refactored it while adding a couple of new things to it.

Here is a short non-extensive synopsis of the changes you should find:

- Added support for TOML configuration files that can be loaded and used for reading the connection parameters, to make it easier to checkout the code to version control and/or deploy the code without revealing critical database info or keys. (The `ConnectionHandlerWithConfig` procedure helps with this.)

- There is a new `Statement` class that allows building statements and makes the code look more elegant, in my opinion.
- There is now a `ConnectionHandler` class for more abstraction, which lets the user use different DBMSs by making very less changes in the code.
- There are interface-type classes that can be implemented to make an API for other relational databases. This can be useful while implementing the API for PostgreSQL and SQLite.
- The module and submodules have been organized to include relevant classes.
- MySQL feature additions: There are a number of functions that have been added to provide more control over the data returned (like for finding field type information etc.), and in case of errors in execution of the queries, meaningful output of the error from the DB engine is produced. (There are other features that have been added too, please have a look at the changes for details.)
- Added tests (more could be added) for the above features, for the MySQL implementation and for other relevant parts too.

The tests can be run using `mason test`, but please ensure that the `TestDBInit.chpl` file is compiled (using `make test_db_init`) and executed first before the tests are run as this file initializes the database tables for the tests, and that the database credentials/details are also updated in the tests. 
Alternatively, you can also use the `runtests.sh` script to run the tests (if this script is used, it ensures that the above step of compiling and executing `TestDBInit.chpl` is performed before running the tests).

I'd really appreciate feedback on the above and any other suggestions, and I hope that this PR could help in making this library even more useful.

Thanks!

krishnadey30 commented:
You are missing new line at end of many files. Please add that.

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+
+        pragma "no doc"
+        // TODO: Use this function at the beginning of each appropriate function
+        proc checkConnected() {
```
I suggest using `_` in from of each internal method.

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+        var _cptr_row: MYSQL_ROW;
+        var _cptr_fields: c_ptr(MYSQL_FIELD);
+        var _fieldTypes: string;
+        var valid: bool;
```
+nit: `_` before the variable.

krishnadey30 replied:
Do you need `valid`?

souris-dev replied:
No, we don't; thanks for noticing this, I'd forgotten to remove it after changing the implementation a bit. (Sorry for the late addressing, a few comments slipped by under my eyes xD).

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+            var fieldVal: string = createStringWithNewBuffer(__get_mysql_field_val_by_name(this._cptr_row, 
+                                                                                           this._cptr_fields, 
+                                                                                           fieldName.localize().c_str()));
+            return fieldVal: t;
```
What if it can't be typecasted? Shouldn't we throw an error?

souris-dev replied:
Since the `return` uses Chapel's explicit type conversion, an `IllegalArgumentError: bad cast from string <string value> to <type>` will be thrown automatically in such a case.

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+            }
+        }
+    }
+}
```
+nit: new line at the end

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+            this._curRow += 1;
+
+            // init the row:
+            var _row = new MySQLRow(nextRow, this._cptr_fields, true);
```
As mentioned in `MySQLRow` do we even need valid? As I can see you never used `false`

souris-dev replied:
You're right, we don't need that anymore. I'll update this.

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+                throw new Error();
+            }
+
+            else {
+                var counter = 0;
+                var _row = this.fetchone();
+
+                while (_row != nil && counter < this._nRows && counter < howManyRows) {
+                    yield _row;
+                    _row = this.fetchone();
+                    counter += 1;
+                }
+            }
+        }
```
Don't you think there should be an error if `howManyRows` exceeds `nRows`

souris-dev replied:
`howManyRows <= _nRows` is a precondition for the function (should include it in the docs though, thanks for noticing it), and it will return rows only up to the actual number of rows (`_nRows`) in the retrieved dataset even if `howManyRows` exceeds `_nRows`; so maybe an error itself is not necessary (seems too severe in my opinion). Rather, I think it would be better to mention about this in the docstring of the function (saying that if `hawManyRows > actual no. of rows`, the function returns all the rows), and possibly, display a warning if this happens (to aid debugging). 

It should also be kept in mind that the user does not directly have the capability to find the number of rows in the fetched set (unless they use `_nRows` itself). 
The reason I left a check for `howManyRows < 0` though is because this would cause the loop itself to malfunction (it won't execute even once).
How does that sound?

krishnadey30 commented on test/mysql/ConnectionTest.chpl:
```diff
@@ -0,0 +1,85 @@
+    conHandler = ConnectionHandlerWithConfig(MySQLConnection, "TransactionTest.chpl");
+}
+
+proc testAutocommit(test: borrowed Test) throws {
```
I would suggest to break this test.

souris-dev replied:
Valid point, I too was a bit reluctant to make it this long but then forgot to update it. I'll split it.

krishnadey30 commented on test/mysql/ConnectionTest.chpl:
```diff
@@ -0,0 +1,85 @@
+// Tests for other methods are located appropriately
+// TODO: replace "appropriately" in above line with proper locations
+
+UnitTest.main();
```
+nit: new line at the end

krishnadey30 commented on test/mysql/CursorTest.chpl:
```diff
@@ -0,0 +1,125 @@
+    conHandler.close();
+}
+
+UnitTest.main();
```
+nit: new line at the end

krishnadey30 commented on test/mysql/FieldTest.chpl:
```diff
@@ -0,0 +1,38 @@
+    conHandler.close();
+}
+
+UnitTest.main();
```
+nit: new line at the end

krishnadey30 commented on test/mysql/RowTest.chpl:
```diff
@@ -0,0 +1,14 @@
+    
+}
+
+UnitTest.main();
```
+nit: new line at the end

krishnadey30 commented on test/mysql/TestDBInit.chpl:
```diff
@@ -0,0 +1,23 @@
+        cursor.close();
+        conHandler.close();
+    }
+}
```
+nit: new line at the end

krishnadey30 commented on test/mysql/TransactionTest.chpl:
```diff
@@ -0,0 +1,44 @@
+    conHandler.close();
+}
+
+UnitTest.main();
```
+nit: new line at the end

krishnadey30 commented on test/mysql/configs/dbconfig.toml:
```diff
@@ -0,0 +1,6 @@
+dbname = "testdb"
+username = "root"
+password = "password"
+autocommit = true
```
+nit: new line at the end

krishnadey30 commented on test/mysql/configs/dbconfig2.toml:
```diff
@@ -0,0 +1,6 @@
+dbname = "testdb"
+username = "root"
+password = "password"
+autocommit = true
```
+nit: new line at the end

krishnadey30 commented on test/mysql/configs/dbconfig_autoc_false.toml:
```diff
@@ -0,0 +1,6 @@
+dbname = "testdb"
+username = "root"
+password = "password"
+autocommit = false
```
+nit: new line at the end

krishnadey30 commented on test/mysql/configs/dbconfig_autoc_true.toml:
```diff
@@ -0,0 +1,6 @@
+dbname = "testdb"
+username = "root"
+password = "password"
+autocommit = true
```
+nit: new line at the end

krishnadey30 commented on test/mysql/configs/dbconfig_noautoc.toml:
```diff
@@ -0,0 +1,5 @@
+host = "localhost"
+dbname = "testdb"
+username = "root"
+password = "password"
```
+nit: new line at the end

krishnadey30 commented on test/mysql/runtests.sh:
```diff
@@ -0,0 +1,38 @@
+./mysql_row_test
+
+echo "[Info] Running TransactionTest"
+./mysql_transaction_test
```
+nit: new line at the end

souris-dev commented on README.md:
````diff
@@ -89,64 +89,80 @@ On Ubuntu do:
 5. Go to repository folder and compile:
 ```bash
-make mysqlex
+make mysql_ex1
````
A valid point, I think that including the required libraries within the code itself along with the `require` statements should be possible, so that those lengthy compiler options aren't needed; I'm currently looking into it.

souris-dev commented on README.md:
```diff
@@ -89,64 +89,80 @@ On Ubuntu do:
+        cursor.execute(new Statement(createStmt));
+        cursor.execute(new Statement("INSERT INTO CONTACTS VALUES (6, 'B');"));
+
+        var stmt: Statement = new Statement("SELECT * FROM CONTACTS WHERE name = ?1", true);
```
Sure, I'll do that. I did try building the docs once but some little improvements are required (formatting stuff mostly). I'll host it on GitHub pages once I make it presentable enough.

souris-dev commented on Mason.toml:
```diff
@@ -1,6 +1,19 @@
+[dependencies]
+
+[system]
+mysqlclient = "21.1.22"
```
Will do

souris-dev commented on src/DatabaseCommunicator.chpl:
```diff
@@ -0,0 +1,65 @@
+    Initializes a connection handler for the given database management system.
+    The configuration information (like username, etc.) are supplied as a TOML file.
+    */
+    proc ConnectionHandlerWithConfig(type className, configFile: string, tomlFileDbconfigRootElem: string = "dbconfig"): owned ConnectionHandler throws {
```
That looks good, I'll do that.

souris-dev commented on example/mysql_ex1.chpl:
```diff
@@ -0,0 +1,27 @@
+module Main {
```
Sure! It's just a simple example showing a statement is supposed to be executed, and how result can be retrieved. I'll add comments explaining this.

souris-dev commented on src/DatabaseCommunicator.chpl:
```diff
@@ -0,0 +1,65 @@
+    use TOML;
+
+    /*
+    Initializes a connection handler for the given database management system.
```
Valid point, I was planning to detail it but forgot to; will do.

souris-dev commented on src/DatabaseCommunicator.chpl:
```diff
@@ -0,0 +1,65 @@
+
+        const DBCONF_ROOT_ELEM = tomlFileDbconfigRootElem;
+
+        const tomlConfigFile = open(configFile, iomode.r);
+        const configs = parseToml(tomlConfigFile);
+
+        var connParams: string = className.getRequiredConnectionParameters();
+        var connString: string = "";
+
+        // build the connection string by taking vaues from the config file
+        for (i, str) in zip(0.., connParams.split(';')) {
+            var connParamValue = configs![DBCONF_ROOT_ELEM]![str]!.toString();
+            connString = connString + connParamValue.strip("\"") + ";"; 
+            
+            // "due to some reason, the TOML parser returns the string enclosed in quotes
+            // hence the above line has to strip it
+        }
+
+        // eliminate the last ';'
+        connString = connString.strip(";");
```
Sure, can do

souris-dev commented on src/DatabaseCommunicator.chpl:
```diff
@@ -0,0 +1,65 @@
+            autocommit = (configs![DBCONF_ROOT_ELEM]!["autocommit"]!.toString()): bool;
+        }
+        else {
+            autocommit = true;
```
Sure; I added that feature later and forgot to update the comments/docs, will update it.

souris-dev commented on src/DatabaseCommunicator/DatabaseCommunicationObjects/QueryBuilder.chpl:
```diff
@@ -0,0 +1,125 @@
+module QueryBuilder {
```
Now that I think, as per your suggestion, this should probably moved to `DatabaseCommunicator` itself, making the imports for the user of the package something like `use DatabaseCommunicator.QueryBuilder;`.
Does that sound right?

krishnadey30 replied:
Yes, this sounds correct.

souris-dev commented on src/DatabaseCommunicator/DatabaseCommunicationObjects/QueryBuilder.chpl:
```diff
@@ -0,0 +1,125 @@
+            :rtype: bool
+        */
+
+        proc isPlaceholderRemaining(): bool {
```
I am a little conflicted about making this an internal-only method, because an user might want to check if their `Statement` has any placeholders remaining in it. At the same time, I can't think of much use cases for it.
@krishnadey30 What is your opinion on this, should it be made internal or not?

krishnadey30 replied:
I am not sure if there is much use. I would suggest have it internal only for now.

souris-dev commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,52 @@
+use UnitTest;
+use DatabaseCommunicator.DatabaseCommunicationObjects.QueryBuilder;
+
+proc statementTest1(test: borrowed Test) throws {
```
Sure, will do

souris-dev commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,52 @@
+    var st2: Statement = new Statement("SELECT * FROM CONTACTS WHERE id = ?1 AND name = ?2;");
+    st2.setValue(1, 0);
+    st2.setValue(2, "testuser");
+    test.assertTrue(st2.getSubstitutedStatement() == "SELECT * FROM CONTACTS WHERE id = ?1 AND name = ?2;");
```
Will do

souris-dev commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,52 @@
+    test.assertTrue(st2.getSubstitutedStatement() == "SELECT * FROM CONTACTS WHERE id = ?1 AND name = ?2;");
+}
+
+proc statementTest3(test: borrowed Test) throws {
```
Will do

souris-dev commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,52 @@
+    test.assertTrue(st3.getSubstitutedStatement() == "SELECT * FROM CONTACTS WHERE id = 1 AND name = 'testuser' AND has_legs = true;");
+}
+
+proc statementTest4(test: borrowed Test) throws {
```
Will do

souris-dev commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,52 @@
+    st4.setValue(3, true);
+    writeln("[Note] IncompleteStatementError should be thrown in this test.");
+    test.assertTrue(st4.isPlaceholderRemaining());
+    test.assertTrue(st4.getSubstitutedStatement() == "SELECT * FROM CONTACTS WHERE id = 1 AND name = ?2 AND has_legs = true;");
```
Yes, I'll create one; thanks for reminding!

souris-dev commented on Makefile:
```diff
@@ -46,6 +46,35 @@ insert_mysql_ex:
+tests: statements_test test_db_init mysql_connection_test mysql_cursor_test mysql_field_test mysql_row_test mysql_transaction_test
+
+mysql_ex1:
+	chpl -o ./bin/mysql_ex1 ./example/mysql_ex1.chpl ./src/mysql/helpers/mysql_helper.h ./src/mysql/helpers/mysql_helper.c -M ./src -M ./src/mysql -I/usr/include/mysql -L/usr/lib/x86_64-linux-gnu -lmysqlclient -lpthread -lz -lm -lrt -lssl -lcrypto -ldl -lresolv
```
Yes, I believe that it should be possible if all the libraries are included with the `require` statements. I'm working on it.

souris-dev commented on Mason.toml:
```diff
@@ -1,6 +1,19 @@
 [brick]
-name = "Cdo"
+name = "DatabaseCommunicator"
 version = "0.1.0"
```
Got it, what should we have? A bit conflicted about how much to increment, would love suggestions.

krishnadey30 replied:
Let the version be this for now. Once we are ready with other 2 connectors we can follow this [guide](https://semver.org/). Also you can refer [NPM guide](https://docs.npmjs.com/about-semantic-versioning)

ben-albrecht replied:
Given we're still in pre-1.0.0, we're in the "anything goes" stage of semantic versioning. I think an increment to `0.2.0` would be reasonable.

souris-dev commented on Mason.toml:
```diff
@@ -1,6 +1,19 @@
+name = "DatabaseCommunicator"
 version = "0.1.0"
-chplVersion = "1.16.0"
 authors = ["Marcos Cleison Silva Santana <marcoscleison@gmail.com>"]
```
Got it, will add that

souris-dev commented on src/DatabaseCommunicator/DatabaseCommunicationObjects/QueryBuilder.chpl:
```diff
@@ -0,0 +1,125 @@
+            }
+
+            var toBeReplacedWith: string;
+            // TODO: add tests for SQL injection
```
Yes, some tests need to be added for this. I'll add them.

souris-dev commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+            :arg autocommit: whether to enable autocommit (optional, true by default)
+            :type autocommit: bool
+        */
+        proc init(host: string, username: string, database: string, password: string, autocommit: bool = true) {
```
Right, I forgot to remove this. Will do so.

souris-dev commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,677 @@
+            :arg autocommit: whether to turn on autocommit
+            :type autocommit: bool
+        */
+        override proc connect(connectionString: string, autocommit: bool = true) throws {
```
Makes sense.

souris-dev commented:
@krishnadey30 Thanks for reviewing my PR! I've made the recommended changes and addressed the comments, please do have a look whenever you have the time.
@ben-albrecht @marcoscleison @buddha314 I'd really appreciate feedback so do have a look whenever you have time, thanks!

ben-albrecht commented on CONTRIBUTORS.md:
```diff
@@ -0,0 +1,3 @@
+Marcos Cleison Silva Santana
+Ben Albrecht
+Krishna Kumar Dey
```
Shouldn't this include Souris Ash with the addition of this PR?

souris-dev replied:
Sure! xD

ben-albrecht commented on README.md:
````diff
@@ -89,64 +89,80 @@ On Ubuntu do:
-writeln("name = ", row[1] );
-}
+```chapel
+module Main {
````
For a minimal example, I don't think this module declaration is necessary.

ben-albrecht commented on src/DatabaseCommunicator/DatabaseCommunicationObjects/ErrorTypes.chpl:
```diff
@@ -0,0 +1,55 @@
+    connection that is not connected.
+    */
+    class NotConnectedError : Error {
+        proc init() {}
```
Are these inits necessary?

souris-dev replied:
Not really, I'll remove them.

ben-albrecht commented on src/DatabaseCommunicator/DatabaseCommunicationObjects/Implementables.chpl:
```diff
@@ -0,0 +1,68 @@
+    */
+    class IField {
+        proc getFieldType() {} // returns an enum type specific to the implementation
+        proc getFieldIdx(): int(32) {} // TODO: rename to getFieldIndex() ?
```
I also prefer full names as long as it's not obnoxiously long, which I don't find `getFieldIndex()` to be.

ben-albrecht commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,614 @@
+    use Map;
+    use SysCTypes;
+    use MySQLNative;
+    //require "mysql.h";
```
These stray comments  should be removed or annotated to describe why we're keeping them around.

ben-albrecht reviewed (commented):
Thank you for your contribution @souris-dev. I took a glance over the changes and made a few comments. This looks like a big improvement in the codebase.

nit: Chapel generally uses 2-space indentation, but it's not a hard requirement. If we care to follow that in this repo, you might be able to auto-fix this with your IDE or a well-constructed global string replacement.

souris-dev commented:
@ben-albrecht Thanks for the review! I've addressed the comments and made the suggested changes, do have a look when possible. (I've also changed the indentation to 2 spaces for convention.)

ben-albrecht reviewed (approved):
My feedback has been addressed.

@krishnadey30 - I will merge once you approve. Also, have you been able to confirm the tests work on your end before we merge?

@marcoscleison - I understand you are busy with other projects at the moment, but if you get a chance to look over the changes before we merge, any additional feedback is welcome.

Thanks again @souris-dev!

krishnadey30 commented on src/DatabaseCommunicator/QueryBuilder.chpl:
```diff
@@ -0,0 +1,126 @@
+      :type toSubstitue: bool
+    */
+    
+    proc init(statement: string, toSubstitute: bool = false) {
```
I think there should be another constructor for these cases.
```chpl
var st3: Statement = new Statement("SELECT * FROM CONTACTS WHERE id = ?1 AND name = ?2 AND has_legs = ?3;", true, 1, 'testUser', true);
```

souris-dev replied:
There's slight ambiguity that can be caused by this kind of a constructor in my opinion. Consider the following cases:
1. The placeholders do not appear in serial order in the statement. So if we have something like:
```sql
SELECT * FROM CONTACTS WHERE id = ?3 AND name = ?1 AND has_legs = ?2;
```
in which order is the user expected to give the variadic arguments? Should it be in the order the placeholders appear in the statement or in natural order?

2. The placeholders are not in order and skip numbers in the index. So something like this is still valid, but it will be quite ambiguous as to which argument sets which placeholder if we use a constructor with variadic arguments:
```sql
SELECT * FROM CONTACTS WHERE id = ?3 AND name = ?1;
```

I think we can have a default behaviour for such cases if we do decide to go for the constructor with variadic args though. What are your thoughts about this?

souris-dev replied:
Just for clarification regarding the second point, would something like this be considered valid?:

```chapel
var st3: Statement = new Statement("SELECT * FROM CONTACTS WHERE id = ?3 AND name = ?1 AND parmanent_address = ?2 AND current_address = ?2 AND future_address = ?4;", true, "testUser", "Lux", 1);
st3.setValue(4, "home");
```

In other words, the user has to set all the placeholders in the constructor or not set any at all (that's the other constructor), right?

krishnadey30 replied:
I think it's okay to make it valid as users may want to modify the placeholder values later on as well.

krishnadey30 commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,611 @@
+        throw new Error();
+      }
+
+      if (howManyRows > this._nRows) {
+        writeln("[Warning] MySQLCursor.fetchsome(howManyRows) called with howManyRows > _nRows; whole result set will be returned.");
+      }
+
+      else {
+        var counter = 0;
+        var _row = this.fetchone();
+
+        while (_row != nil && counter < this._nRows && counter < howManyRows) {
+          yield _row;
+          _row = this.fetchone();
+          counter += 1;
+        }
+      }
+    }
+
```
Doesn't look like you are returning any row in case of howManyRows > actual number of rows.

souris-dev replied:
Will fix this

krishnadey30 commented on test/StatementsTest.chpl:
```diff
@@ -0,0 +1,53 @@
+use UnitTest;
```
Add a test when setValue is called with wrong placeholder index.

krishnadey30 replied:
Also add a test when some larger index is set before a small one.

souris-dev replied:
Sure, I'll add a test for the first scenario, but it should be kept in mind that there's no way to test an exception currently. Hence, I'll be modifying things to display a warning in case setValue is called on an invalid index (halting seems a little extreme in this case in my opinion), do let me know if that looks good.
I'll add a test for the second scenario you mentioned.

krishnadey30 replied:
Let's have a halt and todo for test.

krishnadey30 replied:
Make sure you just check the halt locally.

souris-dev replied:
Sure, I'll add a halt and test it locally.

krishnadey30 commented on test/mysql/ConnectionTest.chpl:
```diff
@@ -0,0 +1,38 @@
+  conHandler.close();
+}
+
+// TODO: Figure out a way to run all the follwing tests in one go
```
This shouldn't be the case. Let's discuss this offline. This may be a bug in the UnitTest library. 😅

krishnadey30 commented on test/mysql/CursorTest.chpl:
```diff
@@ -0,0 +1,125 @@
+use UnitTest;
+use MySQL;
+
+proc cursorOpenExecuteTest(test: borrowed Test) throws {
+  var conHandler = ConnectionHandler.ConnectionHandlerWithConfig(MySQLConnection, "dbconfig.toml");
+  var cursor = conHandler.cursor();
+
+  cursor.execute(new Statement("SELECT * FROM sample"));
+
+  cursor.close();
+  conHandler.close();
+}
```
There are no assert statements. What are you testing here?

souris-dev replied:
I'd like to test if there cursor opens and executes the statement successfully and if it throws an error on wrong SQL syntax. But there's no way to test an exception yet using the UnitTest library hence this function only opens a cursor and executes a statement for now. I'll add a TODO for this in here.

krishnadey30 replied:
Okay, but can you test the successful ones? If yes, please add.

souris-dev replied:
Alright, I think I can check that by using a variation of SQL statements (like inserting a row and checking for it), or some other way.

krishnadey30 commented on test/mysql/RowTest.chpl:
```diff
@@ -0,0 +1,14 @@
+use DatabaseCommunicator.QueryBuilder;
+use UnitTest;
+use MySQL;
+
+proc getValTest(test: borrowed Test) throws {
+
+}
+
+proc getValAsTypeTest(test: borrowed Test) throws {
+  
+}
+
```
Tests are empty.

krishnadey30 commented on test/mysql/TestDBInit.chpl:
```diff
@@ -0,0 +1,24 @@
+module Main {
```
Are you able to use this method from other files?

souris-dev replied:
I'm a little unclear about what you mean - is about the main() method in this module? If yes, I'll have to see if it is accessible from some other file (although I don't see a purpose for doing that, would help if you could elaborate a bit about it, thanks!)

krishnadey30 replied:
This can be ignored for now.

souris-dev commented on src/mysql/MySQL.chpl:
```diff
@@ -0,0 +1,611 @@
+      }
+      else {
+        mysql_autocommit(this._cptr_mysqlconn, 0);
+        mysql_query(this._cptr_mysqlconn, "SET autocommit = 0;");
```
Indeed, thanks for pointing it out

souris-dev commented on src/DatabaseCommunicator/QueryBuilder.chpl:
```diff
@@ -0,0 +1,168 @@
+      this._findPlaceholderIndices(statement);
+    }
+
+    proc init(statement: string, args...?n) {
```
Some git mishap I think, it shows the argument locally. I've updated it now though.

souris-dev commented:
Thanks for the reviews! I have made the changes suggested by @krishnadey30 and made the PR ready to be merged.

souris-dev commented:
Thanks for the reviews! I've made the changes suggested by @krishnadey30 and made the PR ready to be merged.
